### ISSUE TITLE
h2_task.c Software caused connection abort

### ISSUE THREAD

**Issue by impressio** [icing/mod_h2#143]
Apache/v2.4.27
mod_http2/v1.10.7
nghttp2/v1.23.1

Not able to browser website after update to latest Apache and subsequent switch to mpm_event. Reproducible upon clearing Chrome data and history, close browser, etc.

Debug:
h2_task.c(119): (103)Software caused connection abort.

**Comment by icing** [comment]
Can you describe your setup a bit more in detail?

**Comment by impressio** [comment]
LAMP server with php7.0, php-fpm and fastcgi.

Maybe not related to this topic: experienced memory shortage in a month before the upgrade yesterday.

**Comment by icing** [comment]
Does it work when you use http/1.1?

**Comment by impressio** [comment]
yes, I disabled http2 was temporarily.

**Comment by icing** [comment]
Can you set
```
LogLevel http2:trace2
```
and reproduce it? (If your server is not otherwise busy, produces lots of logs)

If there is confidential data in the logs, do not attach them here. You can mail them to stefan.eissing at greenbytes.de

Thanks!

**Comment by frk7z** [comment]
I don't know if this is the same error described in my report(the issue referenced by ConteZero here), but in my error log using trace2 I see this.

```
[Tue Jul 18 17:47:18.506455 2017] [http2:debug] [pid 2990:tid 139904947496704] h2_conn_io.c(281): (9)Bad file descriptor: [client 192.168.9.69:62727] AH03044: h2_conn_io(26): pass_out brigade 17 bytes
[Tue Jul 18 17:47:18.506459 2017] [http2:debug] [pid 2990:tid 139904947496704] h2_session.c(622): [client 192.168.9.69:62727] AH03457: h2_session(26,BUSY,1): sent shutdown notice
[Tue Jul 18 17:47:18.506461 2017] [http2:trace1] [pid 2990:tid 139904947496704] h2_filter.c(144): [client 192.168.9.69:62727] h2_session(26): read, NONBLOCK_READ, mode=0, readbytes=65536
[Tue Jul 18 17:47:18.506463 2017] [http2:debug] [pid 2990:tid 139904947496704] h2_filter.c(194): (103)Software caused connection abort: [client 192.168.9.69:62727] AH03046: h2_session(26): error reading
[Tue Jul 18 17:47:18.506465 2017] [http2:trace1] [pid 2990:tid 139904947496704] h2_session.c(1532): (103)Software caused connection abort: [client 192.168.9.69:62727] h2_session(26,BUSY,1): input gone
```

It's weird that if I use curl(using cookies for login in wp-admin), all ok, but using browser those errors are showed.

PD: sorry to post here if this is not the same error that mine.

**Comment by icing** [comment]
@frk7z and @impressio what you see logged in DEBUG level is that the processing of a request was aborted. When the HTTP/2 connection is closed, the is signalled to all ongoing requests and you will see these DEBUG log lines. This is how it should be.

The question is then, why - at least in @frk7z case, for @impressio I have not seen other log lines, - has the main connection been closed? The line
```
AH03044: h2_conn_io(26): pass_out brigade 17 bytes
```
is the indication that it was not possible to write data out to the client. 

Is there anything indicated on the browser side? Does not request work or do all fail?

Also, if you see curl working and browsers not, it might indicate that your TLS settings are not what they should be. Browsers have tight restrictions on HTTP/2. On HTTP/1.1 less.

**Comment by frk7z** [comment]
Hi @icing , using browser the page keeps blank and loading.

The file with this behavior is "load-styles.php" as I can see on chrome's inspector. It is the only request that does not even have a request header and keeps loading and loading.

Maybe the issue is with TLS settings in apache, I will check it.

Here some logs I uploaded in the other repo.
[wptest.test-error.log.txt](https://github.com/icing/mod_h2/files/1160002/wptest.test-error.log.txt)

**Comment by icing** [comment]
@frk7z thanks for the log! I see in there that sending the response on stream 0-3 gets stuck, until the main connection times out and the stream is aborted. That is the 'Software caused connection abort.' debug you see.

The question is now why it gets stuck and find a way to reproduce this. Nice puzzle.

In the meantime, you may try if larger write sizes might help mitigate this problem. Use
```
H2TLSWarmUpSize 0
H2TLSCoolDownSecs 0
```
on your server.

**Comment by frk7z** [comment]
I tested it out your config, and here some logs.

The url i tested is "https://wptest.test/wp-admin/load-styles.php?c=0&dir=ltr&load%5B%5D=dashicons,admin-bar,common,forms,admin-menu,dashboard,list-tables,edit,revisions,media,themes,about,nav-menus,widgets,site-icon,&load%5B%5D=l10n,buttons,wp-auth-check&ver=4.8" I think the file that have the problem is "load-styles.php"

With your config 
[url_config-wptest.test-error.log.txt](https://github.com/icing/mod_h2/files/1162578/url_config-wptest.test-error.log.txt)

without it 
[url-wptest.test-error.log.txt](https://github.com/icing/mod_h2/files/1162584/url-wptest.test-error.log.txt)

**Comment by Ashus** [comment]
I believe I have the same problem. It is also regarding minified compressed styles (though the request URI length does not matter). On some virtualhosts it work fine, on other there is a problem.

I suspect it might be related to mod_deflate. I haven't got time to test it thoroughly yet, but here are htaccess files of a cache folder of the two projects - one works with http/2, the other hangs with blank page.

[problem-htaccess.txt](https://github.com/icing/mod_h2/files/1167431/problem-htaccess.txt)
[working-htaccess.txt](https://github.com/icing/mod_h2/files/1167432/working-htaccess.txt)

**Comment by icing** [comment]
@Ashus thanks. That seems to confirm my suspicion that the problem is in handling responses with multiple byte chunks in separate buckets. The gzip content filter in the problematic case points to that. It's also what I see in logs from @frk7z 

No, I only need to find the bug...

**Comment by icing** [comment]
I made [a patch](https://github.com/icing/mod_h2/blob/master/patches/h2_stream_stall-v0.diff) against 1.10.7 that might help here. Could you give this a try?

**Comment by icing** [comment]
If you are uncomfortable with a patch file, let me know. I can also check in the changes.

**Comment by Ashus** [comment]
Yes, I'd appreciate it compiled, I just got home from work, currently I'm trying to set up a virutal environment in order to compile it for Debian amd64, Apache2 2.4.27-2 (https://packages.debian.org/sid/apache2). I don't expect results today.

**Comment by frk7z** [comment]
here the log after compiled with patch, using ubuntu 16.04 with ppa from ondrej
[patched-wptest.test-error.log.txt](https://github.com/icing/mod_h2/files/1175078/patched-wptest.test-error.log.txt)

maybe the issue is when WP uses XHR calls from admin-ajax.php

**Comment by impressio** [comment]
I will post "tail" a clean log tomorrow.

Anyway, have anyone experienced same issue using static websites? Not only Wordpress platform (others, PHP/dotNet)?

**Comment by frk7z** [comment]
As I could see, the issue is only present using WordPress admin dashboard(which uses ajax), without it, all is fine. Maybe other platforms or websites that use ajax have the same issue too.

**Comment by godchuanz** [comment]
I believe I'm facing the same issue with WordPress sites. From my observation, it usually happens when there are many concurrent ajax requests. A few of the large files usually never finish transferring, and the requests time out after the period specified by the `Timeout` setting. The problem goes away once I switch to `Protocol http/1.1`.

Could it also may be related to chunking in `mod_deflate`?

Also possible related: https://serverfault.com/questions/859034/chrome-http2-requests-do-not-finish-for-large-gzipped-files-apache

**Comment by Jan-E** [comment]
> Could it also may be related to chunking in mod_deflate?

I do not experience this issue with PHP 5.6.31 x86 nts as mod_fcgid on Windows. The difference may be that I am using mod_brotli, which became avalailable in Apache 2.4.26.
```
LoadModule filter_module modules/mod_filter.so
LoadModule brotli_module modules/mod_brotli.so
<IfModule brotli_module>
	AddOutputFilterByType BROTLI_COMPRESS;DEFLATE text/html text/plain text/xml
</IfModule>
```
This works OK on a https Wordpress site.

**Comment by icing** [comment]
@godchuanz thanks, left a comment there pointing here.
@frk7z thanks for the log. I think I understood now what the problem is. Can you test the new patch I added to the repo [here](https://github.com/icing/mod_h2/blob/master/patches/h2_stream_stall_2.4.x-v2.diff)?

Thanks!

**Comment by icing** [comment]
Made a new version of the patch [here](https://github.com/icing/mod_h2/blob/master/patches/h2_stream_stall_2.4.x-v3.diff). Hope this addresses now all situations for you.

**Comment by Ashus** [comment]
I can confirm the patch v3 works on even on deflated files.
I compiled it successfully, but I had to modify paths inside the patchfile and remove the reference to h2_version.h.
Thanks.

**Comment by frk7z** [comment]
hmmmm on my side the issue is still there.

here the log
[newpatch_wptest.test-error.log.txt](https://github.com/icing/mod_h2/files/1177959/newpatch_wptest.test-error.log.txt)

**Comment by icing** [comment]
@frk7z this log looks really strange. what OS/mpm is this on? It almost looks as if 2 connections work in the same process with the same thread id? Confused...

**Comment by frk7z** [comment]
Using Ubuntu 16.04, Apache2 with mpm_event from ondrej's repository, php-fpm and ssl self-signed.
Modules mod_proxy, mod_proxy_fcgi, mod_ssl and mod_rewrite are enabled by me, others modules enabled are defaults from that repository.

Here the virtualhost configuration, maybe something is wrong.

```
<VirtualHost *:80>
    ServerName wptest.test
    Redirect / https://wptest.test/
</VirtualHost>

<VirtualHost *:443>
    ServerName wptest.test
    Protocols h2 http/1.1
    H2TLSWarmUpSize 0
    H2TLSCoolDownSecs 0
    DocumentRoot /var/www/wptest.test/htdocs
    DirectoryIndex index.php index.html

    SSLEngine on
    SSLCertificateFile "/etc/apache2/selfcerts/wptest.test.crt"
    SSLCertificateKeyFile "/etc/apache2/selfcerts/wptest.test.key"

    SSLCipherSuite EECDH+ECDSA+AESGCM:EECDH+aRSA+AESGCM:EECDH+ECDSA+SHA256:EECDH+ECDSA+SHA384:EECDH+aRSA+SHA256:EECDH+aR
SA+SHA384:EECDH+AES128:!aNULL:!eNULL:!LOW:!3DES:!MD5:!EXP:!PSK:!SRP:!DSS:!RC4:!EDH-DSS-DES-CBC3-SHA:!EDH-RSA-DES-CBC3-SH
A:!KRB5-DES-CBC3-SHA:!CAMELLIA

    SSLProtocol +TLSv1.2 +TLSv1.1 +TLSv1

    <Directory "/var/www/wptest.test/htdocs">
        AllowOverride All
    </Directory>

    ProxyPassMatch ^/(.*\.php(/.*)?)$ fcgi://127.0.0.1:9000/var/www/wptest.test/htdocs/$1

    LogLevel http2:trace2
    ErrorLog ${APACHE_LOG_DIR}/wptest.test-error.log
    CustomLog ${APACHE_LOG_DIR}/wptest.test-access.log combined
</VirtualHost>
```



It's only a ubuntu vm with apache2, php-fpm and mysql using a vanilla wordpress installation, I think is reproducible. If others don't have any issue with this setup, the problem is only mine.

**Comment by icing** [comment]
@frk7z config looks fine to me. Have to read the log again on a quite moment.

In the meantime, I did [release 1.10.8](https://github.com/icing/mod_h2/releases/tag/v1.10.8) so that we can be all on firm ground again what we are testing with. If you could rerun with this version, I'd appreciate your reporting how it went. Thanks!

**Comment by Ashus** [comment]
I had recompiled with your new tag release and page loading still works fine in all cases.

**Comment by frk7z** [comment]
On my tests I found out mod_deflate is the most possible reason for this issue.
Similar or very much alike as @godchuanz said before at the link he posted.

here the logs using mod_deflate and without it.

[deflate_wptest.test-error.log.txt](https://github.com/icing/mod_h2/files/1185349/deflate_wptest.test-error.log.txt)

[nodeflate_wptest.test-error.log.txt](https://github.com/icing/mod_h2/files/1185350/nodeflate_wptest.test-error.log.txt)

both using 1.10.8 release and your setting
```
H2TLSWarmUpSize 0
H2TLSCoolDownSecs 0
```

not tested with mod_brotli instead of mod_deflate

**Comment by icing** [comment]
Thanks for the logs @frk7z. ```mod_deflate``` is certainly not the _cause_. There is a bug in the h2 response transfer from the workers to the main connection, where - on certain timings  and bucket patterns - the response gets stuck.

Can you give [v1.10.9](https://github.com/icing/mod_h2/releases/tag/v1.10.9) a shot? Thanks!

**Comment by frk7z** [comment]
Sure. The same issue, here the logs.

[wptest.test-error.log.txt](https://github.com/icing/mod_h2/files/1185812/wptest.test-error.log.txt)

**Comment by icing** [comment]
Excellent. Please try [v1.10.10](https://github.com/icing/mod_h2/releases/tag/v1.10.10).

**Comment by frk7z** [comment]
Delightful !! now works, although in the logs still exists the text "Software caused connection abort", but you are the expert! so here the logs.

[wptest.test-error.log.txt](https://github.com/icing/mod_h2/files/1185962/wptest.test-error.log.txt)

**Comment by GerritChan** [comment]
I'm having this issue as well, not sure how I can install the latest patch for mod_http2?

**Comment by icing** [comment]
@frk7z super! The "Software caused connection abort" DEBUG logs you see are during the cleanup when everything has been sent successfully. They are normal, yet unnecessary. I might  address them in the future for style points.

@GerritChan what platform? I am sure there are people who can help you.

**Comment by icing** [comment]
Fixed in release v1.10.10

**Comment by GerritChan** [comment]
@icing Yes, sorry: Ubuntu 16.04, Apache 2.4.27 and PHP7.1-fpm. A pointer in the right direction would be greatly appreciated!

**Comment by icing** [comment]
I am sure @oerdnj will soon update his PPA. If you cannot wait and know how to ```configure``` you may try to get the release and compile it yourself. The README has some advice. Related questions are probably best asked at users@httpd.apache.org

**Comment by oerdnj** [comment]
I have just uploaded `apache2_2.4.27-3.0` based on current Debian's 2.4.27-3 from experimental. It has mod_http2 updated to 1.10.10 and mod_ssl patch to support mod_md (to be packaged yet) to https://launchpad.net/~ondrej/+archive/ubuntu/apache2-qa/+packages

I would appreciate if somebody could test the package if they doesn't break anything before I upload them to general public (just give it couple minutes to build).

**Comment by oerdnj** [comment]
@icing And thanks for the ping, that's helpful.

**Comment by icing** [comment]
the latest mod_md patch would be cool. can make a 2.4.x variant tomorrow.

**Comment by oerdnj** [comment]
The v2 trunk version applies fine with just some fuzz to 2.4.27.

**Comment by frk7z** [comment]
@oerdnj no issues from my side using this packages

```
apache2:
  Installed: 2.4.27-3.0+ubuntu16.04.1+deb.sury.org+2
  Candidate: 2.4.27-3.0+ubuntu16.04.1+deb.sury.org+2
```

**Comment by GerritChan** [comment]
@icing @oerdnj 
Works like a charm, thanks alot!

**Comment by impressio** [comment]
Works, and many thanks :)

Saved a "tail" with "trace2" efter upgrade, can send it if relevant.

**Comment by icing** [comment]
Thanks guys, for verifying and reporting back here. No logs needed, unless something goes ```sproink!``` again. :-D

**Comment by candrews** [comment]
I reported the issue to Apache at https://bz.apache.org/bugzilla/show_bug.cgi?id=61382

**Comment by lgtkaushalya** [comment]
I have updated my Apache Package with 'ppa:ondrej/apache2' on 2017-09-23 rd because I faced the same hanging issue when I enabled HTTP2 earlier.

But now even I enabled HTTP2 it still works on HTTP 1.1 without shifting to http2. I'm not sure whether I need to perform any other setting. I used following instructions.

1. sudo add-apt-repository -y ppa:ondrej/apache2 -y
2. apt-key update
3. apt-get update
4. sudo apt-get --only-upgrade install apache2 -y
5. sudo a2enmod http2
6. sudo a2enmod deflate

My conf file is as follows.

<IfModule mod_ssl.c>
        <VirtualHost _default_:443>
                ServerAdmin webmaster@localhost
                Protocols h2 http/1.1
                DocumentRoot /var/www/html/myapplication/web
                DirectoryIndex index.php

                ....
                ....
                ....

                <Directory /var/www/html >
                     AllowOverride All
                     Order allow,deny
                     allow from all
                </Directory>
        </VirtualHost>
</IfModule>

**Comment by icing** [comment]
If you set

```
LogLevel core:debug
LogLevel ssl:debug
LogLevel http2:debug
```

you should see something like this in your log:

```
[...] [ssl:info] [...] [client 127.0.0.1:63856] AH01964: Connection to child 133 established (server noh2.example.org:443)
[...] [ssl:debug] [...] ssl_engine_kernel.c(2151): [client 127.0.0.1:63856] AH02043: SSL virtual host for servername test.example.org found
[...] [core:debug] [...] protocol.c(2310): [client 127.0.0.1:63856] AH03155: select protocol from h2,http/1.1, choices=h2,http/1.1 for server test.example.org
[...] [core:debug] [...] protocol.c(2354): [client 127.0.0.1:63856] AH03156: select protocol, proposals=h2,http/1.1 preferences=h2,http/1.1 configured=h2,http/1.1
[...] [core:debug] [...] protocol.c(2374): [client 127.0.0.1:63856] AH03157: selected protocol=h2

```

**Comment by lgtkaushalya** [comment]
@icing Thanks for your reply and extremely sorry for the late reply because I was out for the vacation.

I looked at the log and it said it falls back to http 1 because 'mpm_prefork' is used. I think this is a new restriction added which was not there earlier. So I tried to switch back to 'mpm_event' but then it seems I cannot use the 'mod_php' to run php and have to switch to php-fpm. Is there any way I can avoid switching to php-fmp and resolve this issue?

**Comment by icing** [comment]
@lgtkaushalya Yes, with the 2.4.27 release (mod_http2 v1.10.7) prefork can no longer use http2. 

As to the best mpm_event configuration for php, I am not the expert and you should ask/google around for advice.
